### PR TITLE
Fix to Issue#93

### PR DESCRIPTION
**Pull request by erikyao**
## 1. Summary

This PR is a fix to [Issue#93](https://github.com/biothings/myvariant.info/issues/93). Following changes will be made to the gnomAD plugin:

- For v2.1.1 data, combined the parsers in `gnomad_parser_genomes.py` and `gnomad_parser_exomes.py` because they are almost identical except one document key ("gnomad_genome" vs "gnomad_exome"). The resulted new parser were further decomposed to 3 components, `ProfileParser`, `SiteQualityMetricParser`, and `PopulationFrequencyParser`, saved in `gnomad_common_parser.py`
- Differentiated v2.1.1 (genomes and exomes) and v3.1.1 (genomes-only) `GnomadVcfRecordParser` objects to `gnomad_v2_parser.py` and `gnomad_v3_parser.py` respectively, using the 3 component parsers above
- Decomposed the mapping objects according to the fields parsed by the 3 component parsers above
     - Mapping of population frequency fields now can be generated dynamically. (Large-scale copy-paste avoided)
- Reorganized the folder structure of manual uploads because v2.1.1 and v3.1.1 have to co-exist.

## 2. What's new in v3.1.1

Currently we're hosting the gnomAD `v2.1.1`. The latest version available is `v3.1.1`. Summarized from [gnomAD downloads](https://gnomad.broadinstitute.org/downloads) and [What's the difference between gnomAD v2 and v3?](https://gnomad.broadinstitute.org/help/whats-the-difference-between-gnomad-v2-and-v3), the difference between these two (with an extra `v2.1.1 liftover` version) is shown below:

| gnomAD | #genome | #exom    | GRCh37/hg19 | GRCh38/hg38 | variant subset                                                              |
|--------|---------|----------|-------------|-------------|-----------------------------------------------------------------------------|
| v2.1.1 | 15,708  | 125,748  | Yes         | No          | non-neuro, non-cancer, controls-only, and non-TOPMed                        |
| v2.1.1 liftover | 15,708  | 125,748  | No         | Yes          | non-neuro, non-cancer, controls-only, and non-TOPMed                        |
| v3.1.1 | 76,156  | **None** | No          | Yes         | non-cancer, non-neuro, non-v2, non-TOPMed, controls/biobanks, 1KG, and HGDP |

## 3. New Folder Structure

If we use `dump(src="gnomad", path="20210610")`, the following folder structure can be prepared:

```text
datasources
|-- gnomad
    |-- 20210610 (*the `data_folder`)
        |-- 2.1.1
        |   |-- exomes
        |   |   |-- gnomad.exomes.r2.1.1.sites.*.vcf.bgz
        |   |   |-- gnomad.exomes.r2.1.1.sites.*.vcf.bgz.tbi
        |   |   `-- liftover_grch38
        |   |       |-- gnomad.exomes.r2.1.1.sites.*.liftover_grch38.vcf.bgz
        |   |       `-- gnomad.exomes.r2.1.1.sites.*.liftover_grch38.vcf.bgz.tbi
        |   `-- genomes
        |       |-- gnomad.genomes.r2.1.1.sites.*.vcf.bgz
        |       |-- gnomad.genomes.r2.1.1.sites.*.vcf.bgz.tbi
        |       `-- liftover_grch38 (*discarded)
        |           |-- gnomad.genomes.r2.1.1.sites.*.liftover_grch38.vcf.bgz
        |           `-- gnomad.genomes.r2.1.1.sites.*.liftover_grch38.vcf.bgz.tbi
        |-- 3.1.1 (*new)
        |   `-- genomes
        |       |-- gnomad.genomes.v3.1.1.sites.chr*.vcf.bgz
        |       `-- gnomad.genomes.v3.1.1.sites.chr*.vcf.bgz.tbi
```

Note that v3.1.1 data are only for hg38 genome; others sources will remain with v2.1.1:

|doc key     |assembly|gnomAD version|path                         |
|------------|--------|--------------|-----------------------------|
|gnomad_exome| hg19   | v2.1.1       | 2.1.1/exomes                |
|gnomad_exome| hg38   | v2.1.1       | 2.1.1/exomes/liftover_grch38|
|gnomad_genome| hg19   | v2.1.1       | 2.1.1/genomes               |
|gnomad_genome| hg38   | v3.1.1 (*new)| 3.1.1/genomes (*new)        |

## 4. Code Structure of Parsers

Component Parsers:

- `ProfileParser`: read and parse basic variant information like positions, ref/alt alleles, variant types, rsIDs, etc.
- `SiteQualityMetricParser`: read and parse the data in "Site Quality Metrics" panel of a variant as shown in the gnomAD browser
- `PopulationFrequencyParser`: read and parse the data in "Population Frequencies" panel of a variant as shown in the gnomAD browser

The main parser, `GnomadVcfRecordParser`, must receive all 3 component parsers to initiate:

- The common `ProfileParser` works well for both v2.1.1 and v3.1.1. It has no internal states so instances are not needed (class methods do all the job).
- The keys of site quality metrics are different in v2.1.1 and v3.1.1, so we created two concrete `SiteQualityMetricParser` classes respectively. They don't have states neither so no instances are needed.
- The keys of population frequencies are different in v2.1.1 and v3.1.1, so we created two concrete `PopulationFrequencyParser` classes respectively. Because these keys are saved for dynamic mapping generation, two global `PopulationFrequencyParser` instances are created for each version.
- Simply put:
    - `record_parser_v2 = GnomadVcfRecordParser(ProfileParserClass, SiteQualityMetricParserClass_v2, PopulationFrequencyParserObj_v2)`
    - `record_parser_v3 = GnomadVcfRecordParser(ProfileParserClass, SiteQualityMetricParserClass_v3, PopulationFrequencyParserObj_v3)`

### CODE REVIEW

**Review by newgene — APPROVED** [review]
Looks great to me! Please feel free to merge it.
